enterprise/internal/insights/compression/commits.go

```diff
@@ -0,0 +1,157 @@
+package compression
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/keegancsmith/sqlf"
+
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/internal/database/basestore"
+	"github.com/sourcegraph/sourcegraph/internal/database/dbutil"
+	"github.com/sourcegraph/sourcegraph/internal/vcs/git"
+)
+
+type DBCommitStore struct {
+	*basestore.Store
+}
+
+type CommitStore interface {
+	Save(ctx context.Context, id api.RepoID, commit *git.Commit) error
+	Get(ctx context.Context, id api.RepoID, start time.Time, end time.Time) ([]CommitStamp, error)
+	GetMetadata(ctx context.Context, id api.RepoID) (CommitIndexMetadata, error)
+	UpsertMetadataStamp(ctx context.Context, id api.RepoID) (CommitIndexMetadata, error)
+	InsertCommits(ctx context.Context, id api.RepoID, commits []*git.Commit) error
+}
+
+func NewCommitStore(db dbutil.DB) *DBCommitStore {
+	return &DBCommitStore{
+		Store: basestore.NewWithDB(db, sql.TxOptions{}),
+	}
+}
+
+func (c *DBCommitStore) With(other basestore.ShareableStore) *DBCommitStore {
+	return &DBCommitStore{Store: c.Store.With(other)}
+}
+
+func (c *DBCommitStore) Transact(ctx context.Context) (*DBCommitStore, error) {
+	txBase, err := c.Store.Transact(ctx)
+	return &DBCommitStore{Store: txBase}, err
+}
+
+func (c *DBCommitStore) Save(ctx context.Context, id api.RepoID, commit *git.Commit) error {
+	ref := commit.ID
+	err := c.Exec(ctx, sqlf.Sprintf(insertCommitIndexStr, id, ref, commit.Committer.Date))
+	if err != nil {
+		return fmt.Errorf("error saving commit for repo_id: %v ref %v: %w", id, ref, err)
+	}
+
+	return nil
+}
+
+func (c *DBCommitStore) InsertCommits(ctx context.Context, id api.RepoID, commits []*git.Commit) error {
+	tx, err := c.Transact(ctx)
+	if err != nil {
+		return err
+	}
+
+	defer tx.Store.Done(err)
```

Comment:
```suggestion
	defer func() { err = tx.Done(err) }()
```

The base store `Done` method works like a filter. If you give it a non-nil error value, it will rollback the current transaction. If you give it a nil error value, it will commit it. In either case, if an error occurs finalizing the transaction, it will modify the input error and return a multierror containing the original error (if it exists) as well as the commit/rollback error. Re-assigning to `err` will modify the value that we're returning from the function.

We put it in a defer block instead of calling it directly so that the `err` value is not captured prior to execution. We want it to operate on the error value _on the way out of the function_.

This can be a tricky idiom so let me know if this makes sense.